### ISSUE TITLE
balance reports: one currency per row, separate currency field

### ISSUE THREAD

**Issue by simonmichael** [simonmichael/hledger#1559]
It is a common request to have the amount quantity and the currency symbol separated in CSV output, rather than combined in the one field, for easier consumption by other tools. As one example, it could be useful to colour code amounts by currency in https://hledger.org/charts.html.

This is problematic in transaction/register reports, where any amount can be a multicommodity amount (especially the running totals in a register report). We don't want to have a varying number of columns, nor to give up the correspondence of rows to transactions or postings.

But it shouldn't be a problem for balance reports (`balance` and friends). There, for an account with a multicommodity balance we could just generate multiple rows, one per commodity. And then we could also put the commodity symbol and amount quantity in separate columns. 

Questions: Any problems with this ? Does it work with multicolumn reports ? Does it work with tree mode ? Should it also be an option for the other output formats, including terminal reports ? Could it help with the problem of too-wide/too-elided multicommodity amounts ? Do it as an option, or always ?

**Comment by simonmichael** [comment]
nMockups: instead of the current `bal -O csv` output:
```
"account","balance"
"Assets:US:ETrade","70.00 GLD, 17.00 ITOT, 5120.50 USD, 36.00 VEA, 294.00 VHT"
"total","70.00 GLD, 17.00 ITOT, 5120.50 USD, 36.00 VEA, 294.00 VHT"
```

it could be:
```
"account","balance"
"Assets:US:ETrade","70.00 GLD"
"Assets:US:ETrade","17.00 ITOT"
"Assets:US:ETrade","5120.50 USD"
"Assets:US:ETrade","36.00 VEA"
"Assets:US:ETrade","294.00 VHT"
"total","70.00 GLD"
"total","17.00 ITOT"
"total","5120.50 USD"
"total","36.00 VEA"
"total","294.00 VHT"
```

or:
```
"account","commodity","amount"
"Assets:US:ETrade","GLD","70.00"
"Assets:US:ETrade","ITOT","17.00"
"Assets:US:ETrade","USD","5120.50"
"Assets:US:ETrade","VEA","36.00"
"Assets:US:ETrade","VHT","294.00"
"total","GLD","70.00"
"total","ITOT","17.00"
"total","USD","5120.50"
"total","VEA","36.00"
"total","VHT","294.00"
```

With multiple periods:
```
"account","2012","2013","2014"
"Assets:US:ETrade","10.00 ITOT, 337.18 USD, 12.00 VEA, 106.00 VHT","70.00 GLD, 18.00 ITOT, -98.12 USD, 10.00 VEA, 18.00 VHT","-11.00 ITOT, 4881.44 USD, 14.00 VEA, 170.00 VHT"
"total","10.00 ITOT, 337.18 USD, 12.00 VEA, 106.00 VHT","70.00 GLD, 18.00 ITOT, -98.12 USD, 10.00 VEA, 18.00 VHT","-11.00 ITOT, 4881.44 USD, 14.00 VEA, 170.00 VHT"
```

it might be:
```
"account","2012 commodity","2012 amount","2013 commodity","2013 amount","2014 commodity","2014 amount"
"Assets:US:ETrade","GLD","0","GLD","70.00","GLD","0"
"Assets:US:ETrade","ITOT","10.00","ITOT","18.00","ITOT","-11.00"
"Assets:US:ETrade","USD","337.18","USD","-98.12","USD","4881.44"
"Assets:US:ETrade","VEA","12.00","VEA","10.00","VEA","14.00"
"Assets:US:ETrade","VHT","106.00","VHT","18.00","VHT","170.00"
"total","GLD","0","GLD","70.00","GLD","0"
"total","ITOT","10.00","ITOT","18.00","ITOT","-11.00"
"total","USD","337.18","USD","-98.12","USD","4881.44"
"total","VEA","12.00","VEA","10.00","VEA","14.00"
"total","VHT","106.00","VHT","18.00","VHT","170.00"
```

**Comment by mjkalyan** [comment]
I'm in favour of separate `account`, `commodity`, and `amount` csv fields (or columns in terminal output) and creating a row for each commodity. i.e. I'd prefer what you describe as `--commodity-column`.

**Comment by simonmichael** [comment]
A downside of moving commodity symbols out of the way is that the numbers all look the same - as if having the same unit, and summable together - which might be deceptive to read at times. But I'd like to try out this feature nevertheless.

**Comment by mjkalyan** [comment]
For CSV output, most will be familiar with bank/exchange-exported CSVs which separate the commodity. Perhaps the terminal output may be confusing? But even the mockup output you posted above for `--commodity-column` is quite legible.
e.g.
```
                  || Commodity    2012    2013     2014 
==================++====================================
 Assets:US:ETrade || GLD             0   70.00        0 
 Assets:US:ETrade || ITOT        10.00   18.00   -11.00 
```

**Comment by Xitian9** [comment]
Could I ask for clarification about whether this is proposed just for CSV output, or are you looking to move all balance reports to this format?

**Comment by simonmichael** [comment]
It was just an idea for the CSV output, but now I think it might have some value for the terminal as well.  (PSA: I added lots of edits to this issue).

**Comment by Xitian9** [comment]
Separating the commodity from the amount it CSV output seems like a pretty straightforward win.

It also has advantages in the text balance report, but comes with disadvantages. If it's a mostly single-commodity report, with only some small instances of multi-commodity amounts, the current behaviour is best.

If it's configurable (i.e. you can keep the old behaviour if you want), I would definitely support introducing an option to control this. I would support either:
- Include the commodity with the amount, as in your first example.
- Include the commodity with the amount by default, but put it in a separate column if specifically requested, as in your second example.

> [--commodity-rows [--commodity-columns=no|yes|one]]

What would `--commodity-columns=one` do that is different from `--commodity-columns=yes`?

> [--show-commodity-in=amount|column|columns]

This seems like a good choice (again, I'm not sure of the difference between column and columns). Not including this option would display a single-line report like current behaviour, `--show-commodity-in=XXX` would use one line per commodity, with the commodity in the specified location.

**Comment by simonmichael** [comment]
> I'm not sure of the difference between column and columns).
> 

=columns would generate multiple commodity columns in a multi-column report - ie wherever we have an amount column now, there'd be a commodity and quantity column. But that's a bit redundant and probably not useful, so instead =column would put the commodity in a single column for the whole row (as in the later mockups).

But I was just brainstorming, and probably less is more. I'd guess just a single flag, --commodity-column or whatever, might be enough.

**Comment by Xitian9** [comment]
The thing I would worry about with a commodity column, is it can expand rapidly if you have a lot of commodities, only a few of which are involved in each column.

```
                  || Commodity    2012    2013     2014 
==================++====================================
 Assets:US:ETrade || GLD             0   70.00        0 
 Assets:US:ETrade || ITOT            0   18.00        0 
 Assets:US:ETrade || USD             0       0     1.00 
 Assets:US:ETrade || GEA         12.00       0    14.00 
 Assets:US:ETrade || VHT        106.00       0        0 
```

As opposed to


```
                  ||       2012        2013       2014 
==================++===================================
 Assets:US:ETrade ||  12.00 GEA   70.00 GLD  14.00 GEA
 Assets:US:ETrade || 106.00 VHT  18.00 ITOT   1.00 USD
```

**Comment by la-wu** [comment]
I realize that the discussion has not been finalized but I thought that this would be a good first exploration of the code base and took a stab at implementing this display change in https://github.com/simonmichael/hledger/pull/1626. I'm certainly amenable to opinions here.

**Comment by simonmichael** [comment]
Implemented by @la-wu's #1626.
